src/middleware/authenticated.ts

```diff
@@ -0,0 +1,46 @@
+import jwt from 'jsonwebtoken';
+import { JWT_SECRET } from '../siteConstants';
+import { DeveloperDocument, Express } from '../types';
+import { createDbConnection, handleCloseConnection } from '../services/database';
+import { developerSchema } from '../models/Developer';
+
+type Payload = {
+  email: string;
+  iat?: number;
+  exp?: number;
+};
+
+export const authenticate: Express.MiddleWare = async (req, res, next) => {
+  let token: string | undefined;
+  // Check if token is set
+  if (req.headers.authorization && req.headers.authorization.startsWith('Bearer')) {
+    [, token] = req.headers.authorization.split(' ');
+  }
+
+  if (!token) {
+    return next(new Error('Unauthorized. Please login to continue.'));
+  }
+
+  let payload: Payload;
+
+  // verify token
+  try {
+    payload = jwt.verify(token, JWT_SECRET) as Payload;
+  } catch (error: unknown | any) {
+    return next(new Error(error.message));
+  }
+
+  // Csheck if developer still exists in the database
+  const connection = createDbConnection();
+  const Developer = connection.model('Developer', developerSchema);
+  const { email } = payload;
+  const currentUser: DeveloperDocument | null = await Developer.findOne({ email });
```

Comment:
I had to annotate `currentUser` because of this:

```
Property 'id' is optional in type 'Document<unknown, any, { createdAt: NativeDate; updatedAt: NativeDate; } & { name: string; apiKey: string; email: string; password: string; usage?: { date: Date; count: number; } | undefined; }> & Omit<...>' but required in type 'DeveloperDocument'.ts(2322)
```

This happens when I save the `currentUser` in the `request`: 
```
  // Grant access
  req.developer = currentUser;
```

Reply:
that means that you'll need to update the DeveloperDocument type to satisify the expected type that's coming from mongoose

Reply:
Done